### ISSUE TITLE
xiaoming90 - Rounding error causing assets to be stolen during withdrawal or lost during deposit

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-06-symmetrical-judging#245]
xiaoming90

medium

# Rounding error causing assets to be stolen during withdrawal or lost during deposit

## Summary

Rounding errors could cause assets to be stolen during withdrawal or lost during deposit under certain conditions.

#### Instance 1 - Token to be stolen during withdrawal

https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/Account/AccountFacetImpl.sol#L27

```solidity
File: AccountFacetImpl.sol
27:     function withdraw(address user, uint256 amount) internal {
28:         AccountStorage.Layout storage accountLayout = AccountStorage.layout();
29:         GlobalAppStorage.Layout storage appLayout = GlobalAppStorage.layout();
30:         require(
31:             block.timestamp >=
32:             accountLayout.withdrawCooldown[msg.sender] + MAStorage.layout().deallocateCooldown,
33:             "AccountFacet: Cooldown hasn't reached"
34:         );
35:         uint256 amountWith18Decimals = (amount * 1e18) /
36:         (10 ** IERC20Metadata(appLayout.collateral).decimals());
37:         accountLayout.balances[msg.sender] -= amountWith18Decimals;
38:         IERC20(appLayout.collateral).safeTransfer(user, amount);
39:     }
```

If the collateral changes to a token with more than 18 decimals in the future, users can drain assets from the contract.

If the collateral's decimals is larger than 18, it will be possible to specify a small `amount,` and the `amountWith18Decimals` will be rounded down to zero in Line 35-36 above. Then, in Line 37, nothing will be deducted from the account as `amountWith18Decimals` is zero.

In Line 38, the `amount`, which is non-zero, of collateral will be transferred to the users.

In summary, non-zero collateral is transferred to users, but nothing is deducted from the account balance. Repeat this process multiple times until all the collaterals in the contract are drained.

Based on the `setCollateral` function, it did not explicitly block collateral with decimals larger than 18. Thus, based on the current design and implementation, it is possible to introduce a collateral with decimals more than 18 in the future.

https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/control/ControlFacet.sol#L95

```solidity
File: ControlFacet.sol
095:     function setCollateral(
096:         address collateral
097:     ) external onlyRole(LibAccessibility.DEFAULT_ADMIN_ROLE) {
098:         GlobalAppStorage.layout().collateral = collateral;
099:         emit SetCollateral(collateral);
100:     }
```

#### Instance 2 - Token to be lost during deposit

The similar issue will occur within the following functions, but with the opposite effect. The users deposit assets to the contract, but their account balance does not increase due to rounding errors.

- `AccountFacetImpl.deposit`
- `AccountFacetImpl.depositForPartyB`
- `AccountFacet.depositAndAllocate`

```solidity
File: AccountFacetImpl.sol
19:     function deposit(address user, uint256 amount) internal {
20:         GlobalAppStorage.Layout storage appLayout = GlobalAppStorage.layout();
21:         IERC20(appLayout.collateral).safeTransferFrom(msg.sender, address(this), amount);
22:         uint256 amountWith18Decimals = (amount * 1e18) /
23:         (10 ** IERC20Metadata(appLayout.collateral).decimals());
24:         AccountStorage.layout().balances[user] += amountWith18Decimals;
25:     }
```

If the user deposits a small number of collateral tokens to the contract, the `amountWith18Decimals` in Line 22 might round down to zero. In this case, the collateral tokens have been transferred to the protocol, but the account balance did not increase in Line 24.

## Impact

Loss of funds if the collateral is updated to a token with decimals more than 18 in the future. Since it requires certain specific conditions for the issue to be exploitable, marking this issue as a Medium instead of a High.

## Code Snippet

https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/Account/AccountFacetImpl.sol#L27

## Tool used

Manual Review

## Recommendation

Consider implementing one of the following fixes to mitigate the issue

#### Solution 1

Update the affected function to revert if rounding errors occur.

```diff
function withdraw(address user, uint256 amount) internal {
	..SNIP..
    uint256 amountWith18Decimals = (amount * 1e18) /
    (10 ** IERC20Metadata(appLayout.collateral).decimals());
+	require(amountWith18Decimals > 0, "Rounding Error Occur")
    accountLayout.balances[msg.sender] -= amountWith18Decimals;
    IERC20(appLayout.collateral).safeTransfer(user, amount);
}
```

```diff
function deposit(address user, uint256 amount) internal {
    GlobalAppStorage.Layout storage appLayout = GlobalAppStorage.layout();
    IERC20(appLayout.collateral).safeTransferFrom(msg.sender, address(this), amount);
    uint256 amountWith18Decimals = (amount * 1e18) /
    (10 ** IERC20Metadata(appLayout.collateral).decimals());
+	require(amountWith18Decimals > 0, "Rounding Error Occur")
    AccountStorage.layout().balances[user] += amountWith18Decimals;
}
```

#### Solution 2

If the protocol does not intend to support collateral tokens with more than 18 decimals, explicitly disallow anyone from configuring such tokens in the future to prevent malicious user from exploiting this issue.

```solidity
function setCollateral(
    address collateral
) external onlyRole(LibAccessibility.DEFAULT_ADMIN_ROLE) {
    GlobalAppStorage.layout().collateral = collateral;
    require(IERC20Metadata(appLayout.collateral).decimals() <= 18, "Token with more than 18 decimals not allowed")
    emit SetCollateral(collateral);
}
```

**Comment by eliteWalrus4164** [comment]
Escalate

The issue is hypothetical:
> If the collateral changes to a token with more than 18 decimals in the future, users can drain assets from the contract.

The contest specified erc20 tokens to interact with protocol:
>Which ERC20 tokens do you expect will interact with the smart contracts?
USDT and USDC

This issue should be considered invalid

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> The issue is hypothetical:
> > If the collateral changes to a token with more than 18 decimals in the future, users can drain assets from the contract.
> 
> The contest specified erc20 tokens to interact with protocol:
> >Which ERC20 tokens do you expect will interact with the smart contracts?
> USDT and USDC
> 
> This issue should be considered invalid
> 

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by securitygrid** [comment]
Future assumption is out of scope.

**Comment by hrishibhat** [comment]
Result:
Low
Unique
Agree with the above escalation

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [SergeKireev](https://github.com/sherlock-audit/2023-06-symmetrical-judging/issues/245/#issuecomment-1651500015): accepted

**Comment by MoonKnightDev** [comment]
Fixed code PR link:
https://github.com/SYMM-IO/symmio-core/pull/18
